pylabnet/scripts/nanopositioners/mcs2_control.py

```diff
@@ -0,0 +1,295 @@
+        (self.step_left, self.step_right, self.walk_left, self.walk_right,
+         self.n_steps, self.is_moving, self.amplitude, self.frequency,
+         self.velocity, self.voltage) = generate_widgets(self.WIDGET_DICT)
+
+        # Additional attributes
+        self.prev_amplitude = [50]*self.NUM_CHANNELS
+        self.prev_frequency = [30]*self.NUM_CHANNELS
+        self.prev_velocity = [100]*self.NUM_CHANNELS
+        self.prev_voltage = [50]*self.NUM_CHANNELS
+        self.voltage_override = False
+
+    def initialize_gui(self):
+        """ Initializes the GUI (assigns channels)"""
+
+        # Iterate through channels
+        for channel_index in range(self.NUM_CHANNELS):
+
+            self._initialize_channel(channel_index)
+
+    def initialize_parameters(self, channel, params):
+        """ Initializes all parameters to values given by params, except for DC voltage
+
+        :param channel: (int) channel index (from 0)
+        :param params: (tuple) params in order n_steps, is_moving, amplitude, frequency, velocity,
+            voltage
+        """
+
+        self.pos.set_parameters(channel, amplitude=params[2])
+        self.pos.set_parameters(channel, frequency=params[3])
+        self.pos.set_parameters(channel, dc_vel=params[4])
+
+        # Measure DC voltage and set it in the GUI
+        self._set_voltage_display(channel)
+
+    def run(self):
+        """ Runs the Positioner control (takes any necessary action) """
+
+        # Iterate through channels
+        for channel_index in range(self.NUM_CHANNELS):
+
+            # Get GUI values
+            params = self.get_GUI_parameters(channel_index)
+
+            # Update current status on GUI
+            self._update_channel(channel_index, params)
+
+            # Handle parameter updates
+            self._update_parameters(channel_index, params)
+
+            # Handle DC change
+            if np.abs(params[5]-self.prev_voltage[channel_index]) > self.DC_TOLERANCE:
+                self.pos.set_voltage(channel_index, params[5])
+
+            # Handle a step event
+            if self.gui.was_button_pressed(self.step_left[channel_index]):
+                self.gui.change_button_background_color(self.step_left[channel_index], color='red')
+                self.pos.n_steps(channel_index, n=-params[0])
+                time.sleep(0.15)
+                self.gui.change_button_background_color(self.step_left[channel_index], color='black')
+                self._set_voltage_display(channel_index)
+            if self.gui.was_button_pressed(self.step_right[channel_index]):
+                self.gui.change_button_background_color(self.step_right[channel_index], color='red')
+                self.pos.n_steps(channel_index, n=params[0])
+                time.sleep(0.15)
+                self.gui.change_button_background_color(self.step_right[channel_index], color='black')
+                self._set_voltage_display(channel_index)
+
+            # Handle walk event
```

Comment:
Is walk the continuous movement we see in GUI?

Reply:
yes